### ISSUE TITLE
Investigate running batch_processing during CI

### ISSUE THREAD

**Issue by Drulex** [neuropoly/spinalcordtoolbox#2888]
Objective: to run `batch_processing.sh` in the CI, preferably only on builds for the `master`branch in order to increase confidence in the software.

We would want:
- To run concurrently as much as possible to avoid increasing the time of the build
- Validate the results (other than exit code) of the script somehow.

TODO:
- [x] How to run? Logic in `.ci.sh` vs build stages?
- [x] When to run? `master` branch, `release` branch?
- [ ] How to validate results of batch processing?

**Comment by Drulex** [comment]
Started to look at this. Initially I added a background job inside the script right after the installation and collected the results at the end:

<details><summary>View diff</summary>
<p>

``` patch
diff --git a/.ci.sh b/.ci.sh
index 883b2b5b..7fd1fa6a 100755
--- a/.ci.sh
+++ b/.ci.sh
@@ -14,6 +14,10 @@ ls -lA bin  # Make sure all binaries and aliases are there
 source python/etc/profile.d/conda.sh  # to be able to call conda
 conda activate venv_sct  # reactivate conda for the pip install below
 
+echo "*** Running batch_processing.sh in the background ***"
+./batch_processing.sh > batch_proc.log 2>&1 &
+bproc_pid=$!
+
 echo *** UNIT TESTS ***
 pytest
 
@@ -30,6 +34,20 @@ echo *** ANALYZE CODE ***
 pip install pylint
 bash -c 'PYTHONPATH="$PWD/scripts:$PWD" pylint -j3 --py3k --output-format=parseable --errors-only $(git ls-tree --name-only -r HEAD | sort | grep -E "(spinalcordtoolbox|scripts|testing).*\.py" | xargs); exit $(((($?&3))!=0))'
 
+echo "Waiting for batch_processing.sh to finish"
+wait $bproc_pid
+bproc_ret=$?
+
+echo "batch_processing.sh done. Results:"
+echo `tail -14 batch_proc.log`
+echo "Return code of batch_processing.sh: ${bproc_ret}"
+
+if [ $bproc_ret -ne 0 ]; then
+  echo "batch_processing.sh exited with non-zero! Full log:"
+  echo `cat batch_proc.log`
+  exit $bproc_ret
+fi
+
 #
 # echo *** BUILD DOCUMENTATION ***
 # pip install sphinx sphinxcontrib.programoutput sphinx_rtd_theme

```

</p>
</details>

While this works ok, it does add run time to the script since `batch_processing.sh` is quite long. Another issue is that managing the concurrent steps of the build from the `.ci.sh` script is not ideal for a few reasons:
- This bash script will grow and become complex as build needs grow
- Some dependencies live at a higher layer. For example if we want to run `batch_processing` only when building master (and not during PRs), there is no way straightforward way to know this from within the `.ci.sh` script (at least from a quick search)
- Need to manage concurrency at script level (no concurrency at the moment)

I looked at Travis's capabilities and this problem is solved using build stages (https://docs.travis-ci.com/user/build-stages/). Build stages would make our build process a lot simpler, basically we would have:
- First stage is installation
- Second stage is running the test scripts (in parallel, managed by travis). This would include unit test, functional tests, batch_processing, linters etc
- Potential third stage executed only for release builds (`release branch`). Would include building docs, deployment stuff etc. To be figured out.

With this scheme we would have one script to run unit tests, one script to run processing, etc. Also we would speed up the current build by parallelizing tests (right now we are running unit tests, functional tests and linting all sequentially, why?).

What do you think @jcohenadad? Was this considered in the past?

@kousu @joshuacwnewton @cfhammill thoughts?

**Comment by Drulex** [comment]
Potential con:
- It is nice to have a `.ci.sh` that contains all build logic and can be run locally in one step.
- Too much `travis` specific logic can make it more difficult to transition to a different CI solution later

**Comment by jcohenadad** [comment]
> With this scheme we would have one script to run unit tests, one script to run processing, etc. Also we would speed up the current build by parallelizing tests (right now we are running unit tests, functional tests and linting all sequentially, why?).
> What do you think @jcohenadad? Was this considered in the past?

nice suggestion. Speed might not necessarily increase though, because AFAIR Travis with free license limits to two simultaneous instances of macOS (so if we add more macOS instances, for unit testing, pylint, etc., they will probably run sequentially). Apart from that, this suggestion makes sense to me, but I might not be aware of all implications.

**Comment by Drulex** [comment]
Stages run within the same job, so assuming a travis job has more than one CPU core available it would speed things things up to setup parallel stages. 

It's not clear how many CPU cores are available for our jobs, but at a quick glance it seems like it's using google compute engine with 2vcpu. This may not yield significant improvements..

**Comment by Drulex** [comment]
Based on an initial assessment it appears that trying to configure `batch_processing` to run in parallel may not yield significant improvements in the current build environment. I propose that for now we simply add it to the end of the `ci` script, configured to run only on the master branch. We can revisit the performance improvements later.

Are you ok with that @jcohenadad? Also what would be a good way to validate the results? If we have a set of "expected" values we could measure relative error and compare against a certain limit? As a first step it's already good to know that it runs without crashing.

**Comment by jcohenadad** [comment]
Yes, I think this is a reasonable approach. And yes, definitely running it on the master branch only is the best decision: it will not hamper our rapid developments on a branch, it will test across more platforms, and once merged on master, CI time does not really matter (ie: it is already merged, so we can focus on something else while Travis is doing its thing). 

Re: how to validate: yes, the set of expected values are presently listed at the end of the script (so far we only do a visual assessment). So that could be automatized. We should also take this opportunity to add more outputs (I can definitely help with that).

**Comment by kousu** [comment]
> I looked at Travis's capabilities and this problem is solved using build stages (https://docs.travis-ci.com/user/build-stages/). Build stages would make our build process a lot simpler, basically we would have:
> 
>     * First stage is installation
> 
>     * Second stage is running the test scripts (in parallel, managed by travis). This would include unit test, functional tests, batch_processing, linters etc
> 
>     * Potential third stage executed only for release builds (`release branch`). Would include building docs, deployment stuff etc. To be figured out.
> 

To clarify a point, you can't really have an "install stage" because --  https://docs.travis-ci.com/user/build-stages/#data-persistence-between-stages-and-jobs

> jobs do not share storage, as each job runs in a fresh VM or container.

So each test run needs to do its own full install. But you can get halfway there by using [caching](https://docs.travis-ci.com/user/build-stages/#warming-up-a-cache-with-expensive-dependencies) or maybe adding an external server for intermediate artifacts. ((Incidentally, Github Actions [provides one](https://github.com/actions/upload-artifact)).

> (right now we are running unit tests, functional tests and linting all sequentially, why?).

Yeah! This seems silly. Split it up!

**Comment by Drulex** [comment]
> To clarify a point, you can't really have an "install stage" because -- https://docs.travis-ci.com/user/build-stages/#data-persistence-between-stages-and-jobs

:facepalm:  Good catch, completely missed this..

**Comment by joshuacwnewton** [comment]
Small obversation I haven't seen mentioned yet: Since `batch_processing.sh` contains 6 separate processing pipelines (t1/t2/t2s/mt/dmri/fmri), could we split up `batch_processing.sh` itself and run each pipeline concurrently?

**Comment by jcohenadad** [comment]
> Since batch_processing.sh contains 6 separate processing pipelines (t1/t2/t2s/mt/dmri/fmri), could we split up batch_processing.sh itself and run each pipeline concurrently?

batch_processing was originally developed as a guideline for our users. The idea is that they would open the script and sequentially run each SCT call line by line, while trying to understand what they do and how they relate to each other. So while the script could in principle be parallelized, it would also introduce confusion for our users. 

Moreover, some inputs (eg warp_* in mt, dwi) are generated by outputs from previous lines (eg register to template for t1). So parallelizing across contrasts is not that trivial.

**Comment by joshuacwnewton** [comment]
Responding to https://github.com/neuropoly/spinalcordtoolbox/issues/2888#issuecomment-694247228:

> Re: how to validate: yes, the set of expected values are presently listed at the end of the script (so far we only do a visual assessment). So that could be automatized. We should also take this opportunity to add more outputs (I can definitely help with that).

This more recent https://github.com/neuropoly/spinalcordtoolbox/issues/3163#issuecomment-759073007 from @kousu/@jcohenadad is relevant:

> Julien doesn't think there's a safe way to automate comparing the numbers, but we can at least move it to cron so that we have a ready point of reference; maybe we can even make up some sort of badge or auto-committed "last output" file, which can then be automatically pulled into the changelog.

What if we still tried to automate comparing the values (and showed this on PRs), yet we don't make it a "Required" check. That way, in some cases it will display an error on CI checks, but it won't halt merging, so we can decide on a per-PR basis whether or not the difference is significant enough to address.

**Comment by joshuacwnewton** [comment]
Something else I came across that might be useful here: The [Upload-Artifact v2](https://github.com/actions/upload-artifact) GitHub Action. 

> This uploads artifacts from your workflow allowing you to share data between jobs and **store data once a workflow is complete.**
>
> See also [download-artifact](https://github.com/actions/download-artifact).

I think we could use this to cache `batch_processing.sh` results between tests and use that as a baseline for comparison.

See also: [GitHub Actions > Guides > Storing workflow data as artifacts](https://docs.github.com/en/actions/guides/storing-workflow-data-as-artifacts#uploading-build-and-test-artifacts).

**Comment by jcohenadad** [comment]
> What if we still tried to automate comparing the values (and showed this on PRs), yet we don't make it a "Required" check. That way, in some cases it will display an error on CI checks, but it won't halt merging, so we can decide on a per-PR basis whether or not the difference is significant enough to address.

Actually, thinking more about this: we could (should) verify the numbers and make it a requirement. As Nick pointed out (i forget when and where), we don't change segmentation/registration/extraction methods very often, and if we do, we can always update the ground truth data. It would feel safer to always verify those numbers.
